### ISSUE TITLE
[Feature] Kotlin support / extensions

### ISSUE THREAD

**Issue by KotlinIsland** [microsoft/playwright-java#1293]
Kotlin is a very popular JVM language, and playwright could expose some Kotlin extensions that augment the existing API.

A small example would be `infix` functions for `or_` and `and_`:
```kt
infix fun Locator.or(other: Locator) = or(other)
infix fun Locator.and(other: Locator) = and(other)

// usage like
println(page.locator("div") or page.locator("span"))
```

Because Kotlin is backwards compatible with Java, it's a possibility that this entire library could be converted to Kotlin, and there would be no disruption to end users using Java (or Scala, or Groovy etc).

- Related #473

**Comment by yury-s** [comment]
Can it be achieved without providing full-fledged Kotlin API?

**Comment by KotlinIsland** [comment]
> Can it be achieved without providing full-fledged Kotlin API?

Yes, because Kotlin is forward and backward compatible with Java, you can include only a set of Kotlin modules within a Java project.

Although, because Kotlin is designed to be so compatible, any Kotlin API can be consumed by Java/Scala/Groovy code, eg:

Here for demonstration purposes we use Kotlin features such as declaration site variance, inheritance via delegation, and extension functions. Then this code is easily interfaced with from Java without any friction at all.
```kt
// utils.kotlin

// unmodifiable collection types, null safe types, declaration site variance, inheritance via delegation
class Foo<T>(val data: MutableList<out T>): List<T> by data

// extension function
fun Locator.doSomething(value: String) { }
```
```java
// Bar.java
public class Bar {
    List<Integer> myList;
    Locator myLocator;

    void something() {
        Foo<Integer> foo = new Foo<Integer>(myList);
        System.out.println(foo.get(1));
        doSomething(myLocator, "hello");
    }
}
```

**Comment by yury-s** [comment]
> Here for demonstration purposes we use Kotlin features such as declaration site variance, inheritance via delegation, and extension functions. Then this code is easily interfaced with from Java without any friction at all.

This code has to be compiled with Kotlin going forward and trying to feed it into javac will fail which essentially means switching playwright to Kotlin or am I missing something?

**Comment by helpermethod** [comment]
I wonder if it wouldn't be better to publish a bunch of Kotlin extension functions as a separate module. Where I would see most benefit by providing options as extension function literals, e.g.

instead of

```kotlin
playwright.chromium().launch(BrowserType.LaunchOptions().setHeadless(false).setSlowMo(500.0))
```

you could write

```kotlin
playwright.chromium().launch { 
    headless = false
    slowMo = 500.0
}
```

**Comment by KotlinIsland** [comment]
> This code has to be compiled with Kotlin going forward and trying to feed it into javac will fail, which essentially means switching Playwright to Kotlin or am I missing something?

Well, in a gradle project, you can take an existing java project, remove the `java` plugin and add the `kotlin` plugin, and `build` would generate the same jvm outputs.

**Comment by twadzins** [comment]
> 
> I guess with a little bit of classpath scanning and something like KotlinPoet you may even be able to generate such extensions functions from the existing API.

Something like KotlinPoet would be nicest, but here's a "typed builder" way to get most of what it sounds like you are looking for. This uses is a new function "options()" which works for any playwright Option argument:

#### Usage of "options() builder"
```kotlin
val browser = playwright.chromium().launch(options {
    // note that command completion of option variables works here
    headless = false 
    args = listOf("--allow-file-access-from-files")
})

```

#### OptionsBuilder.kt
```kotlin
inline fun <reified T> options(noinline init: T.() -> Unit): T? {
    if (init == NOOP) {
        return null
    }
    val options = T::class.java.getDeclaredConstructor().newInstance()
    return options.apply { this@apply.init() }
}

object NOOP : (Any) -> Unit {
    override fun invoke(p1: Any) {
        throw RuntimeException("This line should never be reached under normal use")
    }
}
```
I use the NOOP for creating extension functions with an optional argument for the "Options" object builder) like the following (though there might be a cleaner way to deal with the no-op situation):
```kotlin
fun Locator.shouldBeVisible(init: IsVisibleOptions.() -> Unit = NOOP) = assertThat(this).isVisible(options(init))

//Usage: 
    @Test
    fun `show shouldBeVisible example`() {
        //given
        val page = ...

        //when
        page.navigate("https://www.google.com")

        //then
        // (with no options passed in to shouldBeVisible() )
        page.getByRole(BUTTON, options { name = "I'm Feeling Lucky" }).shouldBeVisible()

        // or (with options passed in to shouldBeVisible() )
        page.getByRole(BUTTON, options { name = "I'm Feeling Lucky" }).shouldBeVisible { timeout = 2000.0 }

        // or (introducing another extension function, which would likely live in some supporting file)
        fun Page.getButtonByName(buttonName: String) = getByRole(BUTTON, options { name = buttonName })
        
        page.getButtonByName("I'm Feeling Lucky").shouldBeVisible()    
    }


```

**Comment by lppedd** [comment]
For the folks using Kotlin: how's not having nullability information?  
Is it fine or does it cause issues on the long run?

I have a Kotlin & TS monorepo, and going the Kotlin route would be nice.  
However TS' types seem to be more precise and probably offer a better devex.

**Comment by lubomsky** [comment]
> For the folks using Kotlin: how's not having nullability information? Is it fine or does it cause issues on the long run?
> 
> I have a Kotlin & TS monorepo, and going the Kotlin route would be nice. However TS' types seem to be more precise and probably offer a better devex.

`how's not having nullability information?` 
Can you give an example that you are concerned about?

I have never encountered any problems related to nulls (with any type ended with `!` like `String!` or `Locator!`).

For example:
```
page.setContent("some example html without head title element")
page.title()
```
will return empty string, not null.

And working with built-in assertion:
```
assertThat(page).hasTitle("some title")
```

will fail like:
```
org.opentest4j.AssertionFailedError: Page title expected to be: some title
Received: 

Call log:
  - Locator.expect with timeout 5000ms
  - waiting for locator(":root")
    9 × locator resolved to <html lang="en">…</html>
      - unexpected value ""
Expected :some title
Actual   :
```
(I do not use any of `PlaywrightAssertions` at all, but I just ran it to check its behavior)

**Comment by hakanai** [comment]
> For the folks using Kotlin: how's not having nullability information? Is it fine or does it cause issues on the long run?

Here's a particularly fun example of `null` treatment in the API.

```
    /**
     * Emulates <a
     * href="https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme">prefers-colors-scheme</a> media
     * feature, supported values are {@code "light"} and {@code "dark"}. See {@link com.microsoft.playwright.Page#emulateMedia
     * Page.emulateMedia()} for more details. Passing {@code null} resets emulation to system defaults. Defaults to {@code
     * "light"}.
     */
    public Optional<ColorScheme> colorScheme;
```

There's a general rule that if you have an `Optional<T>` parameter, that the `Optional` itself cannot be `null`, but here we have a method which actually explicitly states that you _can_ pass `null`. And not only that, you will get different behaviour too.

It's even funnier in this particular case, because `ColorScheme` also has `ColorScheme.NO_PREFERENCE`, which is also a default.

And when you actually read the code, that value starts out as `null`, and there is no sign of it actually defaulting to `ColorScheme.LIGHT`. Presumably that decision is made by the server and the actual default is technically unknown by the client? In which case it would be nicer to have an API to fetch the default settings.

It would be nice if default options were a bit more explicit about where they were coming from. Initially I thought that would just mean adding `UseDefault` and/or `UseSystemDefault` to the enums, but then I discovered some cases which weren't enums which could do with similar treatment, like this one:

```kotlin
    public Optional<ViewportSize> viewportSize;
```

Perhaps instead of `Optional`, all options should use something _like_ `Optional<T>`, but with both `Option.UseStableDefault` and `Option.UseSystemDefault` or similar, so you know exactly where you're getting that default from. (And then make sure every single field in every single options class defaults to `UseStableDefault`, probably easy to automate a test for that.)
